### DISCUSSION TITLE
Assigning textures to a compressed glb doesn't "work" (Babylon)

### DISCUSSION THREAD

**Discussion by RegitGit** [zeux/meshoptimizer#305] [Q&A] [answered]
Hello,

firstly, nice work! The difference between original and compressed size is tremendous! I'm facing an issue though. Namely when using the compressed glb in Babylon and I assign a texture to it the mesh is rendered black. I'm hoping you know what this could be causing, like maybe UVs are getting messed up and there is a option to prevent it that I missed testing.

If I used the same glb with textures already assigned and compress it using your software everythings fine. But I don't want to have textures on the mesh from the get go, that's why I'm loading textures additionally and then assign them to the meshes.
This might be a bit Babylon specific and I don't know if you can help me with that, but I gotta ask.

Thank you!

**Comment by zeux** [comment, marked as answer]
This is somewhat intentional - gltfpack needs to know that the material in question has textures to preserve UV coordinates for the mesh. I'd consider assigning dummy 1x1 textures in the materials for textures that you will need to replace.

**Reply by RegitGit** [reply]
Hi,

I tried it with the 1x1 texture and now Babylon says the mesh at least still has a UV set. Without this texture it doesn't preserve it.  I also tried it with a blank texture of the same size as the original texture, no difference. The problem however is still occuring. It seems to be rooted somewhere else. A difference between the working glb vs the compressed one is when I show the meshes wireframe. I don't know if that might have something to do with it though:
Original
![1](https://user-images.githubusercontent.com/34124165/122735684-8e005c80-d27f-11eb-841f-85d181cba8aa.png)
Compressed (shows Wireframe in the background)
![2](https://user-images.githubusercontent.com/34124165/122735741-9ce70f00-d27f-11eb-8d9f-54f6a0aed69d.png)

The mesh in question is the robotic arm thingie with Kuka written on it (when the texture works ;D)

**Reply by RegitGit** [reply]
New finding: it probably still is an UV issue. Although the mesh now still has a UV set unlike before, the UVs seem to be off.
Original looks like this:
![3](https://user-images.githubusercontent.com/34124165/122738639-75de0c80-d282-11eb-9171-0b74c1618d96.png)
Compressed:
![4](https://user-images.githubusercontent.com/34124165/122738670-7d051a80-d282-11eb-8a95-a06abcb4a777.png)

I also tested with the original texture set. Although it is displayed correctly and it says it has a UV set, when I try to look at the UVs through debug none are displayed.

**Reply by zeux** [reply]
I'm not exactly sure how Babylon works, but when you replace textures and/or materials on compressed assets, it's important to maintain texture offset / scale settings. In Three.js for example these are stored on a texture object, so they need to be copied from the original texture during replacement.

**Reply by zeux** [reply]
Looks like Babylon might be the same and stores these in uOffset / uScale / vOffset / vScale. So you need to copy these from the source texture you're replacing.

**Reply by RegitGit** [reply]
It works! I can update the dummy textures url to my wanted one preserving the offset and scale and it works as intended! I could probably also create a new texture and copy the values as you mentioned, because these were indeed the culprits!

Thank you very much for your help! 😄
